source/opt/insert_extract_elim.cpp

```diff
@@ -63,6 +77,124 @@ bool InsertExtractElimPass::IsVectorType(uint32_t typeId) {
+      // If undefined, give up
+      // TODO(greg-lunarg): Return OpUndef
+      if (shuffle_idx == 0xFFFFFFFF)
+        break;
+      if (shuffle_idx < v1_len) {
+        cid = v1_id;
+        (*pExtIndices)[extOffset] = shuffle_idx;
+      } else {
+        cid = cinst->GetSingleWordInOperand(kVectorShuffleVec2IdInIdx);
+        (*pExtIndices)[extOffset] = shuffle_idx - v1_len;
+      }
+    } else if (cinst->opcode() == SpvOpExtInst &&
+        cinst->GetSingleWordInOperand(kExtInstSetIdInIdx) ==
+          get_module()->GetExtInstImportId("GLSL.std.450") &&
+        cinst->GetSingleWordInOperand(kExtInstInstructionInIdx) ==
+          GLSLstd450FMix) {
+      // If mixing value component is 0 or 1 we just match with x or y.
+      // Otherwise give up.
+      uint32_t comp_idx = (*pExtIndices)[extOffset];
+      std::vector<uint32_t> aIndices = { comp_idx };
+      uint32_t a_id = cinst->GetSingleWordInOperand(kFMixAIdInIdx);
+      ir::Instruction* a_inst = get_def_use_mgr()->GetDef(a_id);
+      uint32_t a_comp_id = DoExtract(a_inst, &aIndices, 0);
+      if (a_comp_id == 0)
+        break;
+      ir::Instruction* a_comp_inst = get_def_use_mgr()->GetDef(a_comp_id);
+      if (a_comp_inst->opcode() != SpvOpConstant)
+        break;
+      uint32_t u = a_comp_inst->GetSingleWordInOperand(kConstantValueInIdx);
+      float* fp = reinterpret_cast<float*>(&u);
+      if (*fp == 0.0)
+        cid = cinst->GetSingleWordInOperand(kFMixXIdInIdx);
+      else if (*fp == 1.0)
+        cid = cinst->GetSingleWordInOperand(kFMixYIdInIdx);
+      else
+        break;
+    }
+    else {
+      break;
+    }
+    cinst = get_def_use_mgr()->GetDef(cid);
+  }
+  // If search ended with CompositeConstruct or ConstantComposite
+  // and the extract has one index, return the appropriate component.
+  // TODO(greg-lunarg): Handle multiple-indices, ConstantNull, special
+  // vector composition, and additional CompositeInsert.
+  if (replId == 0 &&
+      (cinst->opcode() == SpvOpCompositeConstruct ||
+        cinst->opcode() == SpvOpConstantComposite) &&
+      (*pExtIndices).size() - extOffset == 1) {
+    uint32_t compIdx = (*pExtIndices)[extOffset];
+    // If a vector CompositeConstruct we make sure all preceding
+    // components are of component type (not vector composition).
+    uint32_t ctype_id = cinst->type_id();
+    ir::Instruction* ctype_inst = get_def_use_mgr()->GetDef(ctype_id);
+    if (ctype_inst->opcode() == SpvOpTypeVector &&
+        cinst->opcode() == SpvOpConstantComposite) {
+      uint32_t vec_comp_type_id =
+        ctype_inst->GetSingleWordInOperand(kTypeVectorCompTypeIdInIdx);
+      if (compIdx < cinst->NumInOperands()) {
+        uint32_t i = 0;
+        for (; i <= compIdx; i++) {
+          uint32_t compId = cinst->GetSingleWordInOperand(i);
+          ir::Instruction* componentInst = get_def_use_mgr()->GetDef(compId);
+          if (componentInst->type_id() != vec_comp_type_id)
+            break;
```

Comment:
Is this case that is being looked for legal?  If something is of type vector everything must be the same type.  I guess I am  unclear of what "vector composition" means.

Reply:
This chunk of code was designed to handle both OpCompositeConstruct and OpConstantComposite. They are identical except that OpCompositeConstruct allows one additional construction: if constructing a vector, the constituents can be vectors with the same component type as the result of the OpCompositeConstruct. For the moment we are just bailing when we see that, but technically this can be optimized through.